### PR TITLE
Clarify docs on reset method 

### LINKED ISSUE #766: Metric Cache seems to get inconsistent when state variable is a list
## 🐛 Bug

I debugged into this for a longer time now. I am not 100% sure if it is a bug, but i have no idea what i am doing wrong.

I have attached my custom metric to the `Trainer` and call `trainer.validate(model, datamodule)` to compute the metric on the whole Validation set. In each step i print my metric in the `.compute()` function. 

What confused me a lot is, that the `val_mymetric` value is always the same as the metric value of the last batch. Heres my output (only two batches):

```python
mymetric after 1. batch: 0.109375
mymetric after 2. batch: 0.20454546809196472
# end output of .validate():
{'val_cross_entropy': 0.2601596713066101,
 'val_mymetric': 0.20454546809196472}
```

I also found that in the last computation the metric only holds the values from the last batch in `preds`

### My Metric (only important functions)

```python
class MyMetric(Metric, metaclass=abc.ABCMeta):
    def __init__(self):
        super(MyMetric, self).__init__()
        self.softmax = nn.Softmax(dim=1)
        self.add_state("preds", list())
        self.add_state("labels", list())

    def reset(self):
        self.preds.clear()
        self.labels.clear()

    def update(self, preds: torch.Tensor, target: torch.Tensor):
        self.preds.append(self.softmax(preds))
        self.labels.append(target)
```

IMHO the problem is the following:

```python
# metric.py

# save context before switch
cache = {attr: getattr(self, attr) for attr in self._defaults}
```

The cache is taking the exact same object that get's cleared by the `.reset()` function. I think the cache should be built with a copy of each object? The `id()` of the cache object and metric object was the same.

### To Reproduce

as the metric is confidential i am not able to add a bigger code snippet here.

### Expected behavior

At the end of the epoch, the Cache-Content should hold all elements that come in by `update()` and not be cleared by the `.reset()`

### Environment

PyTorch Version (e.g., 1.0): 1.10.0
Flash Version: 0.6.0
OS (e.g., Linux): Ubuntu on Docker
How you installed PyTorch (conda, pip, source): pip with virtualenv
Build command you used (if compiling from source):
Python version: 3.8.12
CUDA/cuDNN version: 11.4
torchmetrics: 0.6.2

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #766
Make documentation more clear that ``reset`` should not be implemented by user.

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/pages/implement.rst [hunk 1]
```diff
@@ -10,11 +10,11 @@ To implement your own custom metric, subclass the base :class:`~torchmetrics.Met
 - ``update()``: Any code needed to update the state given any inputs to the metric.
 - ``compute()``: Computes a final value from the state of the metric.
 
-All you need to do is call ``add_state`` correctly to implement a custom metric with DDP.
-``reset()`` is called on metric state variables added using ``add_state()``.
-
-To see how metric states are synchronized across distributed processes, refer to ``add_state()`` docs
-from the base ``Metric`` class.
+We provide the remaining interface, such as ``reset()`` that will make sure to correctly reset all metric
```
Comment (Borda) [inline docs/source/pages/implement.rst:13]: I would rather move reset to the list above as it is public API and with the updated text you say it is quite an important method...

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
